packages/hardhat/contracts/Pool.sol

```diff
@@ -4,44 +4,70 @@ pragma solidity >=0.6.0 <0.9.0;
 import '@openzeppelin/contracts/token/ERC20/IERC20.sol';
 import '@openzeppelin/contracts/token/ERC20/SafeERC20.sol';
 import '@uniswap/v3-core/contracts/libraries/LowGasSafeMath.sol';
+import '@uniswap/v3-core/contracts/libraries/FullMath.sol';
 
 contract Pool {
   using SafeERC20 for IERC20;
   using LowGasSafeMath for uint256;
 
   IERC20 token;
 
-  uint256 totalBalance = 0;
-  mapping (address => uint) balances;
-
-  uint256 totalShare = 0;
-  mapping (address => uint) shares;
+  uint256 public balance = 0;
+  uint256 public totalShare = 0;
+  mapping (address => uint256) shares;
 
   constructor (address _token) {
     token = IERC20(_token);
   }
 
-  function deposit(uint amount) external {
+  function deposit(uint256 amount) external returns (uint256 share) {
+
+    require(amount > 0, "Pool: deposit amount should be greater than zero");
+
+    share = totalShare > 0
+        ? FullMath.mulDiv(amount, totalShare, balance)
+        : amount.mul(1000000);
```

Comment:
what is this magic number doing?

Reply:
I should move it to a constant. It's the initial rate used to calculate the first share when the Pool is empty, and probably it could be higher.

Why I'm using a multiplier? because after depositFee, the same amount will require a lower amount of shares.

Eg
First Deposit, 1 ETH => 1 Share
Deposit Fee 0.05 ETH
Second Deposit, Now 1 ETH => 0.9523... Shares

Reply:
I'd make it `10 ** 18` so it equates a unit of ERC20, I think it may make numbers easier to understand later on

Reply:
I was thinking about this, `amount` is "how much ERC20 do you want to deposit", which means that it already has at least 18 digits. If I change this constant to something like 10¹⁸, the result will be a number of at least 36 digits.

Reply:
why do you need to multiply it at all then?

Reply:
ERC20 uses 18 decimals but it can be changed. If I don't multiply and we uses a ERC20 with... let's say 6 decimals, that won't be enough and we probably will have issues rounding numbers when we start depositing fees. The more decimals we have here, more accurate it will be.

I can do a mix `amount.mul(10 ** 18 - (token.decimals))`

Reply:
but won't this number grow with deposits anyway? I'm confused, maybe we should discuss later f2f, in the meanwhile go for it